main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -1645,6 +1645,20 @@ object Typer {
           resultEff <- expectTypeM(expected = Type.Empty, actual = eff, exp.loc)
         } yield (constrs, tpe, resultPur, resultEff)
 
+      case KindedAst.Expression.ParYield(frags, exp, loc) =>
+        val patterns = frags.map(_.pat)
+        val parExps = frags.map(_.exp)
+        for {
+          (constrs, tpe, pur, eff) <- visitExp(exp)
+          patternTypes <- inferPatterns(patterns, root)
+          (parExpConstrs, parExpTypes, parExpPurs, parExpEffs) <- seqM(parExps map visitExp).map(unzip4)
+          _ <- InferMonad.point(patternTypes.zip(parExpTypes).map(ts => unifyTypeM(List(ts._1, ts._2), loc)))
+          actualPurs = Type.mkAnd(pur :: parExpPurs, loc)
+          actualEffs = Type.mkUnion(eff :: parExpEffs, loc)
+          resultPur <- expectTypeM(expected = Type.Pure, actual = actualPurs, loc)
```

Comment:
I think we can get better errors than this by expecting each individual fragment expression to be pure, rather than expecting their `And` to be pure.

Also, do we really need `exp` to be pure?